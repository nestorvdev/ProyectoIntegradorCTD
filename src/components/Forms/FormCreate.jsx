import styles from "./styles.module.css";
import React, { useEffect, useState } from "react";
import { Link } from "react-router-dom";
import hidePassword from "./icons/hidePassword.png";
import { AxiosCreate } from "../../axiosCollection/Forms/AxiosForms"
import Spinner from "../spinner/Spinner";
import { Modal } from 'react-responsive-modal';
import 'react-responsive-modal/styles.css';


function FormCreate({ lastLocation, setActiveCreate, setActiveLogin, setLog, setIniciales, setUserName, setUserSurname }) {
    const [name, setName] = useState({ campo: "", valido: true });
    const [surname, setSurname] = useState({ campo: "", valido: true });
    const [errorFullname, setErrorFullname] = useState({ error: "" })
    const [email, setEmail] = useState({ campo: "", valido: true, error: "" });
    const [password, setPassword] = useState({ campo: "", valido: true, error: "" });
    const [confirmPassword, setConfirmPassword] = useState({ campo: "", valido: true, error: "" });
    const [error, setError] = useState("")
    const [formValido, setFormValido] = useState(false)
    const [modalConfirmIsOpen, setModalConfirmIsOpen] = useState(false)


    // @ts-ignore
    const [loading, setLoading] = useState(false);

    /* const baseUrl = "http://localhost:8080/" */
    useEffect(() => {
        setActiveLogin(false)
        // eslint-disable-next-line react-hooks/exhaustive-deps
        setActiveCreate(true)
    }, []);


    const regEx = /^[a-zA-Z\s]{1,70}$/;

    /*CONTROL DE COMPONENTES MEDIANTE HANDLES */
    const handleChangeName = (event) => {
        setName({ ...name, campo: event.target.value })
    }

    const handleChangeSurname = (e) => {
        setSurname({ ...surname, campo: e.target.value })
    }

    const handleChangeEmail = (event) => {
        setEmail({ ...email, campo: event.target.value })
    }

    const handleChangePassword = (event) => {
        setPassword({ ...password, campo: event.target.value })
    }

    const handleChangeConfirmPassword = (event) => {
        setConfirmPassword({ ...confirmPassword, campo: event.target.value })
    }


    /*VALIDACIONES */
    const validarName = () => {

        if (!name.campo || !regEx.test(name.campo)) {
            setName({ ...name, valido: false, })
            setErrorFullname({ error: "Nombre y/o apellido solo pueden tener letras" })
        } else {
            if (surname.valido) {
                setName({ ...name, valido: true })
                setErrorFullname({ error: "" })
            } else {
                setName({ ...name, valido: true })
            }
        }
    }

    const validarSurname = () => {
        if (!surname.campo || !regEx.test(surname.campo)) {
            setErrorFullname({ error: "Nombre y apellido solo pueden tener letras" })
            setSurname({ ...surname, valido: false })
        } else {
            if (name.valido) {
                setErrorFullname({ error: "" })
                setSurname({ ...surname, valido: true })
            } else {
                setSurname({ ...surname, valido: true })
            }
        }
    }

    const validarEmail = () => {
        if (email.campo.length === 0) {
            setEmail({ ...email, valido: false, error: "El email no puede estar vacio" })
        } else {
            setEmail({ ...email, valido: true, error: "" })
        }
    }

    const validarPassword = () => {
        if (!password.campo || password.campo.length <= 6) {
            setPassword({ ...password, valido: false, error: "La contraseña debe tener más de 6 caracteres" })
        } else {
            setPassword({ ...password, valido: true, error: "" })
        }
    }

    const validarConfirmPassword = () => {
        if (confirmPassword.campo.length === 0) {
            setConfirmPassword({ ...confirmPassword, valido: false, error: "El campo es obligatoria" })
        } else if ((password.campo) !== (confirmPassword.campo)) {
            setConfirmPassword({ ...confirmPassword, valido: false, error: "Las contraseñas deben ser iguales" })
        } else {
            setConfirmPassword({ ...confirmPassword, valido: true, error: "" })
        }
    }

    const sendData = (event) => {
        event.preventDefault();
        validarName()
        validarSurname()
        validarEmail()
        validarPassword()
        validarConfirmPassword()

        if (name.campo && surname.campo && email.campo && password.campo && confirmPassword.campo &&
            name.valido && surname.valido && email.valido && password.valido && confirmPassword.valido) {
            AxiosCreate(name.campo, surname.campo, email.campo, password.campo, setFormValido, setLog, setError, setEmail, setPassword, setLoading, lastLocation,openModalConfirm)
            setLoading(true);
        }
        else {
            setFormValido(false)
        }
    }
    function mostrarContrasena(event) {
        let tipo = document.getElementById(event);

        if (tipo.type === "password") {
            tipo.type = "text";
        } else {
            tipo.type = "password";
        }
    }

    const openModalConfirm = (() => {
        setModalConfirmIsOpen(true)
    })

    const closeModalConfirm = () => {
        setModalConfirmIsOpen(false);
    };


    return (

        <div className={styles.containerPrincipal}>
            <div className={styles.containerForm}>
                {loading ? (
                    <Spinner />
                ) : (<>
                    <h3>Crear cuenta</h3>
                    <form className={styles.formFlex} onSubmit={sendData}>
                        <div className={styles.fullName}>
                            <div className={styles.containerFullname}>
                                <div className={`${styles.inputLabel} `}>
                                    <label htmlFor="name">Nombre</label>
                                    <input type="text" name="name" id="name" value={name.campo} className={!name.valido ? styles.inputError : null} onKeyUp={validarName} onChange={handleChangeName} />
                                </div>
                                <div className={`${styles.apellido} `}>
                                    <label htmlFor="surname">Apellido</label>
                                    <input type="text" name="surname" id="surname" value={surname.campo} className={!surname.valido ? styles.inputError : null} onKeyUp={validarSurname} onChange={handleChangeSurname} />
                                </div>
                            </div>
                            <div className={styles.errorContainer}><p className={styles.error}>{errorFullname.error}</p></div>
                        </div>
                        <div className={`${styles.inputLabel} `}>
                            <label htmlFor="email">Correo electrónico</label>
                            <input type="email" name="email" id="email" value={email.campo} className={!email.valido ? styles.inputError : null} onKeyUp={validarEmail} onChange={handleChangeEmail} />
                            <div className={styles.errorContainer}><p className={styles.error}>{email.error}</p></div>
                        </div>
                        <div className={`${styles.inputLabel} `}>
                            <label htmlFor="password">Contraseña</label>
                            <div className={styles.inputHidePassword}>
                                <input type="password" name="password" id="password" value={password.campo} className={!password.valido ? styles.inputError : null} onKeyUp={validarPassword} onChange={handleChangePassword} />
                                <img src={hidePassword} alt="icon hide password" className={styles.hidePassword} onClick={() => mostrarContrasena("password")} />
                            </div>
                            <div className={styles.errorContainer}><p className={styles.error}>{password.error}</p></div>
                        </div>

                        <div className={`${styles.inputLabel} ${styles.confirmPassword} `}>
                            <label htmlFor="confirm-password">Confirmar contraseña</label>
                            <div className={styles.inputHidePassword}>
                                <input type="password" name="confirm-password" id="confirm-password" value={confirmPassword.campo} className={!confirmPassword.valido ? styles.inputError : null} onKeyUp={validarConfirmPassword} onChange={handleChangeConfirmPassword} />
                                <img src={hidePassword} alt="icon hide password" className={styles.hidePassword} onClick={() => mostrarContrasena("confirm-password")} />
                            </div>
                            <div className={styles.errorContainer}><p className={styles.error}>{confirmPassword.error}</p></div>
                        </div>

                        {!formValido && <div className={styles.errorContainer}><p className={styles.error}>{error}</p></div>}
                        <div className={`${styles.inputLabel} ${styles.boton}`}>
                            <button type="submit" id="butCreate">Crear cuenta</button>
                            <p>¿Ya tienes una cuenta?<Link to="/login"> Iniciar sesión</Link></p>
                        </div>
                        <Modal open={modalConfirmIsOpen} onClose={closeModalConfirm} center>
                            <div className={styles.modalConfirm}>
                                <p>Se ha enviado un email de verificación.</p>
                                <Link to="/"><button>Continuar</button></Link>
                            </div>
                        </Modal>
                    </form>
                </>)}
            </div>
        </div>

    )
}

export default FormCreate;
