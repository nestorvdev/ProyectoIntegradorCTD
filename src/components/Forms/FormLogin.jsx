import styles from "./styles.module.css"
import { Link } from "react-router-dom"
import { useEffect, useState } from "react";
import hidePassword from "./icons/hidePassword.png"
import { AxiosLogin } from "../../axiosCollection/Forms/AxiosForms"
import IconError from "./icons/iconError.svg"

export default function FormLogin({ lastLocation, bookingWithoutLogin, setLoading, setActiveLogin, setActiveCreate, setLog }) {
    const [email, setEmail] = useState({ campo: "", valido: true, error: "" });
    const [password, setPassword] = useState({ campo: "", valido: true, error: "" });
    const [error, setError] = useState("")
    const [formValido, setFormValido] = useState(false)

    useEffect(() => {
        setActiveCreate(false)
        setActiveLogin(true)
        // eslint-disable-next-line react-hooks/exhaustive-deps
    }, []);

    /*CONTROL DE COMPONENTES MEDIANTE HANDLES */
    const handleChangeEmail = (event) => { setEmail({ ...email, campo: event.target.value }) }

    const handleChangePassword = (event) => { setPassword({ ...password, campo: event.target.value }) }

    const validarEmailNulo = () => {
        if (!email.campo) {
            setEmail({ ...email, valido: false, error: "El email no puede estar vacío." })
        } else {
            setEmail({ ...email, valido: true, error: "" })
        }
    }
    const validarPasswordNulo = () => {
        if (!password.campo) {
            setPassword({ ...password, valido: false, error: "La contraseña no puede estar vacía." })
        } else {
            setPassword({ ...password, valido: true, error: "" })
        }
    }

    const sendData = (event) => {
        event.preventDefault();
        validarEmailNulo()
        validarPasswordNulo()
        if (password.campo && email.campo && password.valido && email.valido) {
            AxiosLogin(email.campo, password.campo, setFormValido, setLog, setError, setEmail, setPassword, setLoading, lastLocation)
        }        
    }

    function mostrarContrasena() {
        let tipo = document.getElementById("password");
        if (tipo.type=== "password") {
            tipo.type = "text"
        } else {
            tipo.type = "password"
        }
    }

    return (
        <div className={styles.containerPrincipal}>
            <div className={styles.containerForm}>
                <div className={`${styles.errorLogin} ${!bookingWithoutLogin ? styles.hide : null}`}>
                    <img src={IconError} alt="icono de error" />
                    <div >Para realizar una reserva, necesitas estar logueado</div>
                </div>
                <h3>Iniciar sesión</h3>
                <form className={`${styles.formFlex} ${styles.login}`} onSubmit={sendData}>
                    <div className={`${styles.inputLabel} `}>
                        <label htmlFor="email">Correo electrónico</label>
                        <input type="email" name="email" id="email" value={email.campo} className={!email.valido ? styles.inputError : null} onChange={handleChangeEmail} onKeyUp={validarEmailNulo} />
                        <div className={styles.errorContainer}><p className={styles.error}>{email.error}</p></div>
                    </div>
                    <div className={`${styles.inputLabel} `}>
                        <label htmlFor="password">Contraseña</label>
                        <div className={styles.inputHidePassword}>
                            <input type="password" name="password" id="password" value={password.campo} className={!password.valido ? styles.inputError : null} onChange={handleChangePassword} onKeyUp={validarPasswordNulo} />
                            <img src={hidePassword} alt="icon hide password" className={`${styles.hidePassword} ${styles.passLogin}`} onClick={mostrarContrasena} />
                        </div>
                        <div className={styles.errorContainer}><p className={styles.error}>{password.error}</p></div>
                    </div>

                    {!formValido && <div className={styles.errorContainer}><p className={styles.error}>{error}</p></div>}
                    <div className={`${styles.inputLabel} ${styles.boton}`}>
                        <button type="submit" id="login">Ingresar</button>
                        <p>¿Aún no tenés cuenta?<Link to="/create"> Registrate</Link></p>
                    </div>
                </form>
            </div>
        </div>
    )
}