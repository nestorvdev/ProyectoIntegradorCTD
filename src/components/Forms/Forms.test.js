import React from "react";
import { shallow } from "enzyme";
import "@testing-library/jest-dom";
import FormCreate from './FormCreate.jsx';
import FormLogin from "./FormLogin.jsx"

import Enzyme from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
import { createSerializer } from 'enzyme-to-json';

Enzyme.configure({ adapter: new Adapter() });
expect.addSnapshotSerializer(createSerializer({ mode: 'deep' }));

describe("Probando el componente <FormCreate/>", () => {
    let wrapper;
    let setActiveCreate = jest.fn(); //Simula la funcion del set
    let setActiveLogin = jest.fn() //Simula la funcion del set
   
    beforeEach(() => {
        wrapper = shallow(<FormCreate setActiveCreate={setActiveCreate} setActiveLogin={setActiveLogin} />)
    });

    it("Deberia mostrar <FormCreate/> correctamente", () => {
        expect(wrapper).toMatchSnapshot();
    });

    it("Deberia retornar una etiqueta h3", () => {
        let node = wrapper.find("h3")
        expect(node.length).toEqual(1)
    });

    it("Deberia retornar un formulario", () => {
        let node = wrapper.find("form")
        expect(node.length).toEqual(1)
    });

    it('Deberia setear el nombre con el valor pasado', () => {
        wrapper.find('input[id="name"]').simulate('change', {
            target: {
                value: 'nuevoNombre',
            },
        });
        expect(wrapper.find('input[id="name"]').prop('value')).toEqual(
            'nuevoNombre',
        );
    });

    it('Deberia setear el apellido con el valor pasado', () => {
        wrapper.find('input[id="surname"]').simulate('change', {
            target: {
                value: 'nuevoApellido',
            },
        });
        expect(wrapper.find('input[id="surname"]').prop('value')).toEqual(
            'nuevoApellido',
        );
    });

    it('Deberia setear el mail con el valor pasado', () => {
        wrapper.find('input[type="email"]').simulate('change', {
            target: {
                value: 'nuevoMail',
            },
        });
        expect(wrapper.find('input[type="email"]').prop('value')).toEqual(
            'nuevoMail',
        );
    });

    it('Deberia setear la contraseña con el valor pasado', () => {
        wrapper.find('input[id="password"]').simulate('change', {
            target: {
                value: 'nuevaContraseña',
            },
        });
        expect(wrapper.find('input[id="password"]').prop('value')).toEqual(
            'nuevaContraseña',
        );
    });

    it('Deberia setear confirmar contraseña con el valor pasado', () => {
        wrapper.find('input[id="confirm-password"]').simulate('change', {
            target: {
                value: 'nuevaContraseña',
            },
        });
        expect(wrapper.find('input[id="confirm-password"]').prop('value')).toEqual(
            'nuevaContraseña',
        );
    });
    it("Deberia traer cinco parrafos de error", ()=>{
        expect(wrapper.find("p.error").length).toEqual(5)
    });

    it("Deberia llamar a event.preventDefault cuando se envia el formulario", ()=>{
        const event = { preventDefault: () => {} }
        jest.spyOn(event, 'preventDefault')
        wrapper.find('form').simulate('submit', event)
                
        expect(event.preventDefault).toBeCalled()
    })


    /*it('Deberia arrojar un mensaje de error por enviar la contraseña con menos de 6 caracteres', () => {
        wrapper.find('input[id="name"]').simulate('change', {
            target: {
                value: 'nuevoNombre',
            },
        });
        
        wrapper.find('input[id="surname"]').simulate('change', {
            target: {
                value: 'nuevoApellido',
            },
        });
        
        wrapper.find('input[type="email"]').simulate('change', {
            target: {
                value: 'nuevoMail@mail.com',
            },
        });
        wrapper.find('input[id="password"]').simulate('change', {
            target: {
                value: '12345',
            },
        });
        wrapper.find('input[id="confirm-password"]').simulate('change', {
            target: {
                value: '1234567',
            },
        });
        wrapper.find("button").simulate("click")
        expect(wrapper.find("div.errorContainer:nth-child(4)").text()).toEqual("")
    });*/

    it('Deberia llamarse la funcion setActiveCreate', () => {
        expect(setActiveCreate).toHaveBeenCalled()
    });

});

describe("Probando el componente <FormLogin/>", () => {
    let wrapper;
    let setActiveCreate = jest.fn(); //Simula la funcion del set
    let setActiveLogin = jest.fn() //Simula la funcion del set
    let setLog = jest.fn()

    beforeEach(() => {
        wrapper = shallow(<FormLogin setActiveCreate={setActiveCreate} setActiveLogin={setActiveLogin} setLog={setLog}/>)
    })

    afterEach(() => {
        jest.clearAllMocks();
    });

    it("Deberia mostrar <FormLogin/> correctamente", () => {
        expect(wrapper).toMatchSnapshot();
    })

    it("Deberia retornar una etiqueta h3", () => {
        let node = wrapper.find("h3")
        expect(node.length).toEqual(1)
    });

    it("Deberia retornar un formulario", () => {
        let node = wrapper.find("form")
        expect(node.length).toEqual(1)
    });

    it('Deberia setear el mail con el valor pasado', () => {
        wrapper.find('input[type="email"]').simulate('change', {
            target: {
                value: 'nuevoMail',
            },
        });
        expect(wrapper.find('input[type="email"]').prop('value')).toEqual(
            'nuevoMail',
        );
    });

    it('Deberia setear la contraseña con el valor pasado', () => {
        wrapper.find('input[type="password"]').simulate('change', {
            target: {
                value: 'nuevaContraseña',
            },
        });
        expect(wrapper.find('input[type="password"]').prop('value')).toEqual(
            'nuevaContraseña',
        );
    });

    it("Deberia traer tres parrafos de error", ()=>{
        expect(wrapper.find("p.error").length).toEqual(3)
    });

    it("Deberia llamar a event.preventDefault cuando se envia el formulario", ()=>{
        const event = { preventDefault: () => {} }
        jest.spyOn(event, 'preventDefault')
        wrapper.find('form').simulate('submit', event)
                
        expect(event.preventDefault).toBeCalled()
    })

});
