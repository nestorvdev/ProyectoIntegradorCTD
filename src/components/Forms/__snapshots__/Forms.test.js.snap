// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Probando el componente <FormCreate/> Deberia mostrar <FormCreate/> correctamente 1`] = `
<div
  className="containerPrincipal"
>
  <div
    className="containerForm"
  >
    <h3>
      Crear cuenta
    </h3>
    <form
      className="formFlex"
      onSubmit={[Function]}
    >
      <div
        className="fullName"
      >
        <div
          className="containerFullname"
        >
          <div
            className="inputLabel "
          >
            <label
              htmlFor="name"
            >
              Nombre
            </label>
            <input
              className={null}
              id="name"
              name="name"
              onChange={[Function]}
              onKeyUp={[Function]}
              type="text"
              value=""
            />
          </div>
          <div
            className="apellido "
          >
            <label
              htmlFor="surname"
            >
              Apellido
            </label>
            <input
              className={null}
              id="surname"
              name="surname"
              onChange={[Function]}
              onKeyUp={[Function]}
              type="text"
              value=""
            />
          </div>
        </div>
        <div
          className="errorContainer"
        >
          <p
            className="error"
          />
        </div>
      </div>
      <div
        className="inputLabel "
      >
        <label
          htmlFor="email"
        >
          Correo electrónico
        </label>
        <input
          className={null}
          id="email"
          name="email"
          onChange={[Function]}
          onKeyUp={[Function]}
          type="email"
          value=""
        />
        <div
          className="errorContainer"
        >
          <p
            className="error"
          />
        </div>
      </div>
      <div
        className="inputLabel "
      >
        <label
          htmlFor="password"
        >
          Contraseña
        </label>
        <div
          className="inputHidePassword"
        >
          <input
            className={null}
            id="password"
            name="password"
            onChange={[Function]}
            onKeyUp={[Function]}
            type="password"
            value=""
          />
          <img
            alt="icon hide password"
            className="hidePassword"
            onClick={[Function]}
            src="hidePassword.png"
          />
        </div>
        <div
          className="errorContainer"
        >
          <p
            className="error"
          />
        </div>
      </div>
      <div
        className="inputLabel confirmPassword "
      >
        <label
          htmlFor="confirm-password"
        >
          Confirmar contraseña
        </label>
        <div
          className="inputHidePassword"
        >
          <input
            className={null}
            id="confirm-password"
            name="confirm-password"
            onChange={[Function]}
            onKeyUp={[Function]}
            type="password"
            value=""
          />
          <img
            alt="icon hide password"
            className="hidePassword"
            onClick={[Function]}
            src="hidePassword.png"
          />
        </div>
        <div
          className="errorContainer"
        >
          <p
            className="error"
          />
        </div>
      </div>
      <div
        className="errorContainer"
      >
        <p
          className="error"
        />
      </div>
      <div
        className="inputLabel boton"
      >
        <button
          id="butCreate"
          type="submit"
        >
          Crear cuenta
        </button>
        <p>
          ¿Ya tienes una cuenta?
          <Link
            to="/login"
          >
             Iniciar sesión
          </Link>
        </p>
      </div>
      <ForwardRef
        center={true}
        onClose={[Function]}
        open={false}
      >
        <div
          className="modalConfirm"
        >
          <p>
            Se ha enviado un email de verificación.
          </p>
          <Link
            to="/"
          >
            <button>
              Continuar
            </button>
          </Link>
        </div>
      </ForwardRef>
    </form>
  </div>
</div>
`;

exports[`Probando el componente <FormLogin/> Deberia mostrar <FormLogin/> correctamente 1`] = `
<div
  className="containerPrincipal"
>
  <div
    className="containerForm"
  >
    <div
      className="errorLogin hide"
    >
      <img
        alt="icono de error"
        src="iconError.svg"
      />
      <div>
        Para realizar una reserva, necesitas estar logueado
      </div>
    </div>
    <h3>
      Iniciar sesión
    </h3>
    <form
      className="formFlex login"
      onSubmit={[Function]}
    >
      <div
        className="inputLabel "
      >
        <label
          htmlFor="email"
        >
          Correo electrónico
        </label>
        <input
          className={null}
          id="email"
          name="email"
          onChange={[Function]}
          onKeyUp={[Function]}
          type="email"
          value=""
        />
        <div
          className="errorContainer"
        >
          <p
            className="error"
          />
        </div>
      </div>
      <div
        className="inputLabel "
      >
        <label
          htmlFor="password"
        >
          Contraseña
        </label>
        <div
          className="inputHidePassword"
        >
          <input
            className={null}
            id="password"
            name="password"
            onChange={[Function]}
            onKeyUp={[Function]}
            type="password"
            value=""
          />
          <img
            alt="icon hide password"
            className="hidePassword passLogin"
            onClick={[Function]}
            src="hidePassword.png"
          />
        </div>
        <div
          className="errorContainer"
        >
          <p
            className="error"
          />
        </div>
      </div>
      <div
        className="errorContainer"
      >
        <p
          className="error"
        />
      </div>
      <div
        className="inputLabel boton"
      >
        <button
          id="login"
          type="submit"
        >
          Ingresar
        </button>
        <p>
          ¿Aún no tenés cuenta?
          <Link
            to="/create"
          >
             Registrate
          </Link>
        </p>
      </div>
    </form>
  </div>
</div>
`;
